### PR TITLE
Developer templates json

### PR DESCRIPTION
**Pull request by Freymaurer**
Initializes the basic logic for Template and Template json parsing

### CODE REVIEW

**Comment by Freymaurer** [comment]
@HLWeil @kMutagene Someone of you solved different expected line endings?

image

File: src/ARCtrl/Templates/Templates.Json.fs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    let decode = Decode.Auto.generateDecoder<CompositeCell>()
+
+module CompositeHeader =
+
+    let encode = Encode.Auto.generateEncoder<CompositeHeader>()
+
+    let decode = Decode.Auto.generateDecoder<CompositeHeader>()
+
+module ArcTable =
```
Comment (HLWeil) [inline src/ARCtrl/Templates/Templates.Json.fs:29]: Good to have! And I think you cood move the base-type converters to ISA.Json. 

But should we save the templates as `ISA-json`?

File: src/ARCtrl/Templates/Templates.Json.fs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            let valueDecoder = CompositeCell.decode
+            let decodedValues = get.Required.Field "values" (Decode.map' keyDecoder valueDecoder) |> System.Collections.Generic.Dictionary
+            ArcTable.create(
+                get.Required.Field "name" Decode.string,
+                decodedHeader,
+                decodedValues
+            )
+        )
+
+module Json =
```
Comment (HLWeil) [inline src/ARCtrl/Templates/Templates.Json.fs:55]: Why this break in module naming?

File: src/ARCtrl/Templates/Templates.fs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    static member init(templateName: string) =
+        let guid = System.Guid.NewGuid()
+        let table = ArcTable.init(templateName)
+        Template(guid, table, templateName)
+
+    member this.SemVer
+        with get() = ARCtrl.SemVer.SemVer.tryOfString this.Version
+
+    member this.structurallyEquivalent (other: Template) =
```
Comment (HLWeil) [inline src/ARCtrl/Templates/Templates.fs:48]: Capitalize first letter

File: src/ISA/ISA/ArcTypes/ArcTable.fs [hunk 4]
```diff
@@ -621,20 +621,17 @@ type ArcTable =
         ]
         |> String.concat "\n"
 
+    member this.structurallyEquivalent (other: ArcTable) =
```
Comment (HLWeil) [inline src/ISA/ISA/ArcTypes/ArcTable.fs:624]: Capitalize first letter

File: src/ISA/ISA/ArcTypes/ArcTable.fs [hunk 5]
```diff
@@ -621,20 +621,17 @@ type ArcTable =
...
+        let n = this.Name = other.Name
+        let headers = Aux.compareSeq this.Headers other.Headers
+        let values = Aux.compareSeq (this.Values |> Seq.sortBy (fun x -> x.Key)) (other.Values |> Seq.sortBy (fun x -> x.Key))
+        n && headers && values
+
     // custom check
     override this.Equals other =
```
Comment (HLWeil) [inline src/ISA/ISA/ArcTypes/ArcTable.fs:631]: Add identity Equivalence as alternative, explicit method.

https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/how-to-test-for-reference-equality-identity

File: tests/ARCtrl/SemVer.Tests.fs
Comment (HLWeil) [inline tests/ARCtrl/SemVer.Tests.fs:1]: <3

File: tests/ARCtrl/Templates.Tests.fs
Comment (HLWeil) [inline tests/ARCtrl/Templates.Tests.fs:1]: Rename to Templates.Json.Tests?

File: src/ARCtrl/SemVer.fs
[the remark this answers is not in the document]
Comment (HLWeil) [reply src/ARCtrl/SemVer.fs:1]: Not sure either though. Will there be other uses for it?

**Comment by HLWeil** [comment]
> @HLWeil @kMutagene Someone of you solved different expected line endings?
> 
> image

Weird. Not sure what happened here and why it didn't fail previously.

File: src/ARCtrl/SemVer.fs
[the remark this answers is not in the document]
Comment (Freymaurer) [reply src/ARCtrl/SemVer.fs:1]: Might be relevant for ARC information, if we want to version arcs in SemVer. Will keep it there for now

File: src/ARCtrl/Templates/Templates.Json.fs (hunk 1 above)
Comment (Freymaurer) [reply src/ARCtrl/Templates/Templates.Json.fs:29]: I do not have any hard feelings on where to put the template logic. We can move it to ISA and ISA.Json.

File: src/ARCtrl/Templates/Templates.Json.fs (hunk 2 above)
Comment (Freymaurer) [reply src/ARCtrl/Templates/Templates.Json.fs:55]: Ups, thats a leftover from my testfile. Fixed!

File: src/ARCtrl/Templates/Templates.Json.fs (hunk 1 above)
Comment (HLWeil) [reply src/ARCtrl/Templates/Templates.Json.fs:29]: Ah sry a bit of miscommunication. I was referring to two things:
- Move the more general parsers (CompositeCell, ArcTable, ArcAssay... ) to ISA.Json
- Then the question whether we should store the templates as objects specified in [isa-json](https://isa-specs.readthedocs.io/en/latest/isajson.html), i.e. as [`processes`](https://isa-specs.readthedocs.io/en/latest/isajson.html#process-schema-json)

File: src/ARCtrl/SemVer.fs
[the remark this answers is not in the document]
Comment (HLWeil) [reply src/ARCtrl/SemVer.fs:1]: Fine with me

File: src/ISA/ISA/ArcTypes/ArcTable.fs (hunk 5 above)
Comment (Freymaurer) [reply src/ISA/ISA/ArcTypes/ArcTable.fs:631]: Done! 👍

File: src/ISA/ISA/ArcTypes/ArcTable.fs [hunk 6]
```diff
@@ -621,25 +615,30 @@ type ArcTable =
         ]
         |> String.concat "\n"
 
+    member this.StructurallyEquivalent (other: ArcTable) =
```
Comment (HLWeil) [inline src/ISA/ISA/ArcTypes/ArcTable.fs:618]: Now that I think about it. Maybe rename to "StructurallyEquals" for consistent naming?

File: src/ISA/ISA/ArcTypes/IdentifierSetters.fs [hunk 7]
```diff
@@ -2,6 +2,11 @@
 
 open ARCtrl.ISA.Identifier
 
+let setArcTableName (newName: string) (table: ArcTable) =
```
Comment (HLWeil) [inline src/ISA/ISA/ArcTypes/IdentifierSetters.fs:5]: Why not "ArcTable.setName"?

File: tests/ARCtrl/ArcTable.fs
Comment (HLWeil) [inline tests/ARCtrl/ArcTable.fs:1]: Are those the copied TestObjects from ISA.Tests?
Maybe we could just reference theses underlying Test projects from ARCtrl tests?

File: tests/ARCtrl/InvestigationFile.fs
Comment (HLWeil) [inline tests/ARCtrl/InvestigationFile.fs:1]: Same here, maybe just add the reference.
Could also be nice for top level tests like arc read and write

**Review by HLWeil — CHANGES_REQUESTED** [review]

File: src/ISA/ISA/ArcTypes/IdentifierSetters.fs (hunk 7 above)
Comment (Freymaurer) [reply src/ISA/ISA/ArcTypes/IdentifierSetters.fs:5]: This must be unique right? So we only allow updating this unsafe. ArcTables implements a rename option

File: tests/ARCtrl/ArcTable.fs
Comment (Freymaurer) [reply tests/ARCtrl/ArcTable.fs:1]: referercing did not work with the error ~"[<EntryPoint>] can only be used once and must be last in project"

Comment (Freymaurer) [reply tests/ARCtrl/ArcTable.fs:1]: So if you want this, we would need a project without tests and just util. Which might be preferrable with all that extra stuff we accumulated

File: tests/ARCtrl/InvestigationFile.fs
Comment (Freymaurer) [reply tests/ARCtrl/InvestigationFile.fs:1]: See other

File: tests/ARCtrl/ArcTable.fs
Comment (HLWeil) [reply tests/ARCtrl/ArcTable.fs:1]: Ah I see, yeah might actually be beneficial. But probably better in an additional PR.

File: src/ISA/ISA/ArcTypes/IdentifierSetters.fs (hunk 7 above)
Comment (HLWeil) [reply src/ISA/ISA/ArcTypes/IdentifierSetters.fs:5]: Ah you mean you wanted to hide the functionality?

**Review by HLWeil — APPROVED** [review]
🚀

File: src/ISA/ISA/ArcTypes/IdentifierSetters.fs (hunk 7 above)
Comment (Freymaurer) [reply src/ISA/ISA/ArcTypes/IdentifierSetters.fs:5]: Exactly as the other unique identifier setters

Comment (Freymaurer) [reply src/ISA/ISA/ArcTypes/IdentifierSetters.fs:5]: We have a visible member for this, but only on higher level types where we can sanitycheck the changes

**Pull request merged**